### PR TITLE
Stable sort for CPU

### LINKED ISSUE #38681: Stable torch.sort and torch.argsort
## 🚀 Feature
Add stable version of `torch.sort` and `torch.argsort`. Stable sort algorithms sort repeated elements in the same order that they appear in the input.

## Motivation

In some applications we need to preserve the order of equal elements while sorting.
## Alternatives

Now we can do:

```
def stable_argsort(arr, dim=-1, descending=False):
    arr_np = arr.detach().cpu().numpy()
    if descending:
        indices = np.argsort(-arr_np, axis=dim, kind='stable')
    else:
        indices = np.argsort(arr_np, axis=dim, kind='stable')
    return torch.from_numpy(indices).long().to(arr.device)
```

and

```
def stable_sort(arr, dim=-1, descending=False):
    arr_np = arr.detach().cpu().numpy()
    if descending:
        indices = np.sort(-arr_np, axis=dim, kind='stable')
    else:
        indices = np.sort(arr_np, axis=dim, kind='stable')
    return torch.from_numpy(indices).as_type(arr)
```

But I think that it will be cool to have this functionality in `torch.sort` and `torch.argsort`.

### PR DESCRIPTION
**Pull request by nikitaved**
Fixes [#38681](https://github.com/pytorch/pytorch/issues/38681) for the CPU.

### CODE REVIEW

**Comment by nikitaved** [comment]
`argsort` does belong to the legacy functionality. Does it actually make sense to extend it as well?

File: torch/_torch_docs.py
Comment (glaringlee) [inline torch/_torch_docs.py:7643]: Change the doc string if we agree on CUDA throw

**Comment by glaringlee** [comment]
@ailzhang This break the xla build due to signature change in sorting functions. Can you advise?

**Comment by glaringlee** [comment]
@nikitaved `pytorch_linux_backward_compatibility_check_test` is broken since the signature of the sort function changed, please add the changed function into the following list to bypass them, expiration date 01/31/2021 should be fine.
https://github.com/pytorch/pytorch/blob/master/test/backward_compatibility/check_backward_compatibility.py#L53

Here is a quick link to the compatibility check error log.
https://app.circleci.com/pipelines/github/pytorch/pytorch/256847/workflows/dfb6662c-2cf4-45f4-85a9-840b3547bb34/jobs/9959226

**Comment by JackCaoG** [comment]
@glaringlee I will draft a pr to pt/xla to address the upstream signature change

File: aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp
Comment (nikitaved) [reply aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp:706]: Yes, but TORCH_CHECK will break the tests from the `common_method_invokations.py`.

**Comment by glaringlee** [comment]
> `argsort` does belong to the legacy functionality. Does it actually make sense to extend it as well?

This is what we current does for argsort, it seems to me that no extra work is needed.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LegacyDefinitions.cpp#L28

File: aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp
Comment (glaringlee) [reply aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp:706]: I found similar usage here:
https://github.com/pytorch/pytorch/blob/5d2639f8b808fe3c15755b65b8993a6bab8293cf/torch/testing/_internal/common_methods_invocations.py#L1957

Try this: 
`('sort', (S, M, S), (0, False, True), 'stable', (), NO_ARGS, [onlyCPU])`

And then, at the beginning of the file import onlyCPU like this
```
from torch.testing._internal.common_device_type import \
    (skipCUDAIfNoMagma, skipCPUIfNoLapack, skipCPUIfNoMkl, skipCUDAIfRocm,
     expectedAlertNondeterministic, precisionOverride, onlyCPU)
```
If it works, you can apply this to all the `stable` tests. Please add a comment about the reason why you skip CUDA

**Comment by glaringlee** [comment]
@nikitaved 
test_torch.py need to apply the same thing, adding onlyCPU for stable sort one.
Here is an example:
https://github.com/pytorch/pytorch/blob/20a388241ad8e64208c152db243aaa7a439c9564/test/test_torch.py#L6704

**Comment by glaringlee** [comment]
@JackCaoG @ailzhang This is ready to land. Should I land this code first or you guys land xla code first? thx

**Comment by glaringlee** [comment]
@nikitaved Can you please rebase the code to resolve the conflict in check_backwrard_compatibility.py?

**Comment by JackCaoG** [comment]
> @JackCaoG @ailzhang This is ready to land. Should I land this code first or you guys land xla code first? thx

@glaringlee you can land this one first, I will merge the xla one once this is merged

**Review by glaringlee — APPROVED** [review]
LGTM now.

**Comment by glaringlee** [comment]
@nikitaved Hey, I got a landing failure due to conflict on file common_methods_invocations.py, can you rebase again? Sry about this.

**Comment by glaringlee** [comment]
@nikitaved sry for reopen this. This code passed all our CI tests, but breaks the fb internal model tests, torch script can not decode the previous saved model file with the new sort function signature. One way is to put all the `stable = False` after `*` in the argument list, but it will make it harder to write generic tests. (Since now you have to specify `stable=`  all the time). 
I'll see whether there is a better way to solve this problem, will update here tomorrow.

**Comment by glaringlee** [comment]
@JackCaoG 
Hey Jack, our CI xla test is still broken after your back out, can you check on your side if this is something related?
Here is an error log:
https://app.circleci.com/pipelines/github/pytorch/pytorch/261850/workflows/30fd09da-7110-4776-bfa1-90511721c8a4/jobs/10259174

**Comment by JackCaoG** [comment]
> @JackCaoG
> Hey Jack, our CI xla test is still broken after your back out, can you check on your side if this is something related?
> Here is an error log:
> https://app.circleci.com/pipelines/github/pytorch/pytorch/261850/workflows/30fd09da-7110-4776-bfa1-90511721c8a4/jobs/10259174
> pls let me know.

I have reverted the stable sort pr, this failure seems to be a different issue. Will take a look.

**Comment by jysohn23** [comment]
> Here is an error log:
> [app.circleci.com/pipelines/github/pytorch/pytorch/261850/workflows/30fd09da-7110-4776-bfa1-90511721c8a4/jobs/10259174](https://app.circleci.com/pipelines/github/pytorch/pytorch/261850/workflows/30fd09da-7110-4776-bfa1-90511721c8a4/jobs/10259174)
> pls let me know.

We've just merged PR on our end, could you try run against our head again? Thanks.

**Comment by glaringlee** [comment]
> > Here is an error log:
> > [app.circleci.com/pipelines/github/pytorch/pytorch/261850/workflows/30fd09da-7110-4776-bfa1-90511721c8a4/jobs/10259174](https://app.circleci.com/pipelines/github/pytorch/pytorch/261850/workflows/30fd09da-7110-4776-bfa1-90511721c8a4/jobs/10259174)
> > pls let me know.
> 
> We've just merged PR on our end, could you try run against our head again? Thanks.

@ailzhang is looking at this now. @JackCaoG @jysohn23

**Comment by glaringlee** [comment]
@nikitaved, @mruberry has some suggestions on how to implement this without breaking fb internal model test.

File: test/test_sort_and_select.py
Comment (mruberry) [inline test/test_sort_and_select.py:117]: I suggest adding some comparisons with NumPy's stable sort, too.

File: test/test_sort_and_select.py
Comment (mruberry) [inline test/test_sort_and_select.py:119]: Use the `@dtypes` decorator

File: test/test_torch.py
Comment (mruberry) [inline test/test_torch.py:6823]: Instead of updating these tests, implement an OpInfo for sort and remove them. Ping me if you have questions about how to do this.

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2469]: Let's implement an OpInfo for sort and remove all these method_test entries.

**Comment by mruberry** [comment]
Hey @nikitaved!

Here's the issue I think we're dealing with.

Internally with Facebook we have tools for authoring models and services for running models. Services for running models are sometimes not updated at the exact same moment that the tools for authoring models are, and this causes the authoring tool to produce a signature that the service can't understand.

Luckily this isn't too hard to deal with. PyTorch matches signatures in the order they appear in native_functions.yaml, so we can do the following:

- keep sort's current signatures in native_functions.yaml
- add new signatures BELOW those signatures with the keyword argument with NO DEFAULT VALUE

That way, when a PyTorch program calls sort without the new kwarg they'll match the previous signatures.

The document should still reflect that the kwarg has a default value, even though the implementation doesn't make that immediately obvious. The default behavior is implemented by the old signatures.

In the future we can cleanup native_functions.yaml by removing the old signatures and adding the default value to the kwarg in the new signatures.

As for testing, there are a few things to do (I made some comments inline):

- compare with NumPy on a larger variety of inputs in test_sort_and_select
- remove the old-style tests in test_torch and common_methods_invocations and implement an OpInfo for sort

Ping me if you have any questions, @nikitaved.

File: test/test_torch.py
Comment (nikitaved) [reply test/test_torch.py:6823]: Will do! A small question: Is there a plan for the OpInfo's sample method to support generators? Could be quite handy in implementing several samplers which attach specific use cases and then concatenating them without any additional memory allocation...

Comment (mruberry) [reply test/test_torch.py:6823]: That's a really cool idea that I hadn't thought about. Would you like to open an issue to suggest/discuss it?

**Comment by glaringlee** [comment]
> Hey @nikitaved!
> 
> Here's the issue I think we're dealing with.
> 
> Internally with Facebook we have tools for authoring models and services for running models. Services for running models are sometimes not updated at the exact same moment that the tools for authoring models are, and this causes the authoring tool to produce a signature that the service can't understand.
> 
> Luckily this isn't too hard to deal with. PyTorch matches signatures in the order they appear in native_functions.yaml, so we can do the following:
> 
> * keep sort's current signatures in native_functions.yaml
> * add new signatures BELOW those signatures with the keyword argument with NO DEFAULT VALUE
> 
> That way, when a PyTorch program calls sort without the new kwarg they'll match the previous signatures.
> 
> The document should still reflect that the kwarg has a default value, even though the implementation doesn't make that immediately obvious. The default behavior is implemented by the old signatures.
> 
> In the future we can cleanup native_functions.yaml by removing the old signatures and adding the default value to the kwarg in the new signatures.
> 
> As for testing, there are a few things to do (I made some comments inline):
> 
> * compare with NumPy on a larger variety of inputs in test_sort_and_select
> * remove the old-style tests in test_torch and common_methods_invocations and implement an OpInfo for sort
> 
> Ping me if you have any questions, @nikitaved.

@mruberry Great solution! This completely solved the fb internal build inconsistency for sort operator!

**Comment by nikitaved** [comment]
>     * add new signatures BELOW those signatures with the keyword argument with NO DEFAULT VALUE
> 
> 
> That way, when a PyTorch program calls sort without the new kwarg they'll match the previous signatures.
> 
> The document should still reflect that the kwarg has a default value, even though the implementation doesn't make that immediately obvious. The default behavior is implemented by the old signatures.

Does it mean that I have to also have two overloaded version of `sort` in CPP?

**Comment by mruberry** [comment]
> > ```
> > * add new signatures BELOW those signatures with the keyword argument with NO DEFAULT VALUE
> > ```
> > 
> > 
> > That way, when a PyTorch program calls sort without the new kwarg they'll match the previous signatures.
> > The document should still reflect that the kwarg has a default value, even though the implementation doesn't make that immediately obvious. The default behavior is implemented by the old signatures.
> 
> Thank you, @mruberry . I have a small question. Does it mean that I have to also have two overloaded versions of `sort` in CPP?

Yes, but the version without the kwarg can delegate (call) the version with the kwarg.

File: aten/src/ATen/native/native_functions.yaml
Comment (nikitaved) [inline aten/src/ATen/native/native_functions.yaml:6350]: @mruberry , @glaringlee , are you fine with these schemas? I had to make `bool?`  instead of `bool` because some methods in ATen call to `sort(int)`, and this confuses compiler with the overloaded function selection.

Comment (glaringlee) [reply aten/src/ATen/native/native_functions.yaml:6360]: @mruberry 
This seems fine to me, any advise?

**Comment by glaringlee** [comment]
@nikitaved Can we fix the lint error? I will run fb internal tests for this.

**Comment by nikitaved** [comment]
@glaringlee , done. I will remove all the tests and move them to OpInfo such that this new `kwarg` could be handled properly. Hopefully fb tests are going to be happy now.

**Comment by glaringlee** [comment]
> @glaringlee , done. I will remove all the tests and move them to OpInfo such that this new `kwarg` could be handled properly. Hopefully fb tests are going to be happy now.

@nikitaved, sry to bother, can you make a new PR for this change and past the pr number here? The PR was landed once and reverted. There is a parallel PR in fb internal which is marked closed already, I am not able to import this PR to fb internal anymore.

**Comment by rgommers** [comment]
Completed in gh-51790, closing.

**Pull request closed without merging**
